### PR TITLE
Back Compat Fix for TestMethods with parameters

### PR DESCRIPTION
**Pull request by jayaranigarg**

### CODE REVIEW

File: src/Adapter/MSTest.CoreAdapter/Helpers/ReflectHelper.cs [hunk 1]
```diff
@@ -130,6 +130,43 @@ public bool HasAttributeDerivedFrom(MemberInfo memberInfo, Type baseAttributeTyp
             return false;
         }
 
+        /// <summary>
+        /// Returns true when specified class/member has attribute which implements specific interface.
+        /// </summary>
+        /// <param name="memberInfo">The member info.</param>
+        /// <param name="interfaceType">The interface type.</param>
+        /// <param name="inherit">Should look at inheritance tree.</param>
+        /// <returns>An object derived from Attribute that corresponds to the instance of found attribute.</returns>
```
Comment (singhsarab) [inline src/Adapter/MSTest.CoreAdapter/Helpers/ReflectHelper.cs]: Returns bool

Comment (singhsarab) [reply src/Adapter/MSTest.CoreAdapter/Helpers/ReflectHelper.cs]: Cant we use HasAttributeDerivedFrom ? 
If not, let's rename the method to HasAttributeImplementingInterface/HasAttributeImplementing/ HasAttributeImplementedFrom

File: src/Adapter/MSTest.CoreAdapter/Helpers/ReflectHelper.cs [hunk 2]
```diff
@@ -130,6 +130,43 @@ public bool HasAttributeDerivedFrom(MemberInfo memberInfo, Type baseAttributeTyp
...
+
+            if (interfaceType == null)
+            {
+                throw new ArgumentNullException(nameof(interfaceType));
+            }
+
+            // Get all attributes on the member.
+            Dictionary<string, object> attributes = this.GetAttributes(memberInfo, inherit);
+
+            // Try to find the attribute that is derived from baseAttrType.
```
Comment (singhsarab) [inline src/Adapter/MSTest.CoreAdapter/Helpers/ReflectHelper.cs]: there is no baseAttrType in this context

File: src/Adapter/MSTest.CoreAdapter/Helpers/ReflectHelper.cs [hunk 3]
```diff
@@ -130,6 +130,43 @@ public bool HasAttributeDerivedFrom(MemberInfo memberInfo, Type baseAttributeTyp
...
+            foreach (object attribute in attributes.Values)
+            {
+                Debug.Assert(attribute != null, "ReflectHeler.DoesAttributeImplement: internal error: wrong value in the attrs dictionary.");
```
Comment (singhsarab) [inline src/Adapter/MSTest.CoreAdapter/Helpers/ReflectHelper.cs]: Nit:ReflectHelper*

File: src/Adapter/MSTest.CoreAdapter/Discovery/TestMethodValidator.cs [hunk 4]
```diff
@@ -51,12 +51,15 @@ internal virtual bool IsValidTestMethod(MethodInfo testMethodInfo, Type type, IC
...
 
+            var ignoreParameters = this.reflectHelper.DoesAttributeImplement(testMethodInfo, typeof(ITestDataSource), true);
+
             // Todo: Decide wheter parameter count matters.
             // The isGenericMethod check below id to verify that there are no closed generic methods slipping through.
             // Closed generic methods being GenericMethod<int> and open being GenericMethod<T>.
             var isValidTestMethod = testMethodInfo.IsPublic && !testMethodInfo.IsAbstract && !testMethodInfo.IsStatic
                                     && !testMethodInfo.IsGenericMethod
-                                    && testMethodInfo.IsVoidOrTaskReturnType();
+                                    && testMethodInfo.IsVoidOrTaskReturnType()
```
Comment (singhsarab) [inline src/Adapter/MSTest.CoreAdapter/Discovery/TestMethodValidator.cs:61]: What is the warning message users sees, should we add a test for that ?

Also, if parameters are more than what DataSource needs, what then ?

File: src/Adapter/MSTest.CoreAdapter/Helpers/ReflectHelper.cs (hunk 1 above)
Comment (jayaranigarg) [reply src/Adapter/MSTest.CoreAdapter/Helpers/ReflectHelper.cs]: Done.

File: src/Adapter/MSTest.CoreAdapter/Helpers/ReflectHelper.cs (hunk 2 above)
Comment (jayaranigarg) [reply src/Adapter/MSTest.CoreAdapter/Helpers/ReflectHelper.cs]: Done

File: src/Adapter/MSTest.CoreAdapter/Helpers/ReflectHelper.cs (hunk 3 above)
Comment (jayaranigarg) [reply src/Adapter/MSTest.CoreAdapter/Helpers/ReflectHelper.cs]: Done

File: src/Adapter/MSTest.CoreAdapter/Discovery/TestMethodValidator.cs (hunk 4 above)
Comment (jayaranigarg) [reply src/Adapter/MSTest.CoreAdapter/Discovery/TestMethodValidator.cs:61]: Added test.

If there is parameter count mismatch, if will be caught while trying to invoke the test Method.

File: src/Adapter/MSTest.CoreAdapter/Helpers/ReflectHelper.cs (hunk 1 above)
Comment (jayaranigarg) [reply src/Adapter/MSTest.CoreAdapter/Helpers/ReflectHelper.cs]: There was a slight variation in the functionality, so I thought keeping a seperate method would be better.

File: src/Adapter/MSTest.CoreAdapter/Helpers/ReflectHelper.cs [hunk 5]
```diff
@@ -130,6 +130,46 @@ public bool HasAttributeDerivedFrom(MemberInfo memberInfo, Type baseAttributeTyp
...
+
+            // Get all attributes on the member.
+            Dictionary<string, object> attributes = this.GetAttributes(memberInfo, inherit);
+
+            if (attributes != null)
+            {
+                // Try to find the attribute that implements interfaceType.
+                foreach (object attribute in attributes.Values)
+                {
+                    Debug.Assert(attribute != null, "ReflectHelper.HasAttributeImplementingInterface: internal error: wrong value in the attrs dictionary.");
```
Comment (acesiddhu) [inline src/Adapter/MSTest.CoreAdapter/Helpers/ReflectHelper.cs:160]: how does Debug.Assert help? also if we hit this condition we will fail any ways in the next line

**Pull request closed without merging**